keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,943 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This Kep proposes adding a feature (`PodsPerNodeAssignmentPolicy`) that provides
+users with an explicit API and scheduler support for 1:1 pod to node assignment.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Customers trying to deploy batch workloads on Kubernetes have filed multiple
+requests (See [1](https://github.com/kubernetes/kubernetes/issues/68827),
+[2](https://github.com/kubernetes/kubernetes/issues/105287) for details) for 1:1
+pod-to-node assignment. The most common use case for enforcing 1:1 pod to node
+assignment is HPC workloads (MPI and Spark).
+
+Currently there are two approaches to achieve this:
+
+-   Set resource requests to something high enough to prevent scheduling any
+    other pods but daemonset and static pods. This is tricky to do and error
+    prone, especially for smaller nodes. This will not be possible if node size
+    is not known in advance.
+-   Use pod anti affinity. Users will need to define the same set of labels on
+    all pods that needs to be repelled, whether or not they actually request 1:1
+    pod to node assignment, it is also expensive to calculate especially for
+    cluster autoscaler.
+
+A simple and native way to force 1:1 pod-to-node assignment is therefore needed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod.
+    -   This is enforced by both scheduler and kubelet.
+    -   This feature is implemented as a filter, and so features associated with
+        filters applies to it like preemption and autoscaling.
+
+### Non-Goals
+
+-   Find the best fit node for a given pod
+-   Avoid pods with `pod.spec.coexistPolicy` field set to
+    `DaemonsetAndStaticPods` from getting preempted.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new pod-level API that allows users to specify how many "worker" pods are
+allowed on a node. The API is implemented by the scheduler as a filter plugin
+and kubelet as an admission predicate.
+
+### User Stories (Optional)
```

Comment:
This feature will be also useful for running databases on Kubernetes. Sometimes databases need various non-namespaced kernel settings tweaks. eg: https://dbexamstudy.blogspot.com/2021/11/is-kubernetes-slowing-down-my-database.html . But those settings might not be right for other applications in general.

You can imagine a DBaaS running natively on Kubernetes which uses 1:1 Pod / Node assignment where Nodes are sized appropriately for said pod's resource requirements.

Reply:
@tamalsaha IIUC, this also needs the ability to choose the target node the pod is going to be scheduled on, which is out of scope of this feature. I think `nodeSelector` (https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) might be more feasible for this feature as it will schedule pods based on node labels. Please let me know if I misunderstood your use case.

@ahg-g wdyt?

Reply:
nodeselectors will be needed for this use-case but also a way to make sure 2 pods don't get assigned to the same node (pod-affinity can be used for that).

Reply:
Correct, we still need to ensure 1:1 even with node affinity/selector, nice use case.

Reply:
I didn't notice earlier that the goal was also to run exclusively on the node. Thanks @tamalsaha for the clarification.

Would the following describe the use case correctly?
```
A user wants a Pod (for stateful workloads) to be scheduled on a node with specific configuration where those settings might not be right for other applications in general. A combination of `nodeSelector`/`node affinity` with 1:1 pod to node assignment can be used to achieve this goal.
```

Reply:
Thanks for summarizing the use-case. LGTM.

Reply:
I would use the same language Tamal used to describe the use case instead of trying to abstract it too much, it is more relatable when the description is concrete.